> @bernhardkaindl FYI -- do you mind trying this fix out?

* The KeyError from old environments is completely gone! :+1: 

When checking out an old spack commit, this is still happening (but we could fix this with PR #25890):

 ```py
(with -d: InvalidDatabaseVersionError: )Expected database version 5 but found version 6.
```

However, `spack reindex` ends execution at the same error. `rm opt/spack/.spack-db/index.json` is needed instead, after which, reindex is done automatically.

With a reindexed db, `"AttributeError: 'str' object has no attribute 'get'"` blocks operation when older spack versions try to read the new `misc_cache`. Like `rm opt/spack/.spack-db/index.json`, a `spack clean -m` triggers a regeneration of the misc_cache, but that takes a long time. Here is a debug trace of that:
```py
spack -d install anyword 2>&1 |sed "s@$SPACK_ROOT@\$spack@;s/2021.*]/time]/"
==> [time] Imported install from built-in commands
==> [time] Reading config file $spack/etc/spack/defaults/config.yaml
==> [time] Imported install from built-in commands
==> [time] Reading config file $spack/etc/spack/defaults/repos.yaml
Traceback (most recent call last):
  File "$spack/bin/spack", line 100, in <module>
    sys.exit(spack.main.main())
  File "$spack/lib/spack/spack/main.py", line 797, in main
    return _invoke_command(command, parser, args, unknown)
  File "$spack/lib/spack/spack/main.py", line 525, in _invoke_command
    return_val = command(parser, args)
  File "$spack/lib/spack/spack/cmd/install.py", line 395, in install
    args.spec, concretize=True, tests=tests)
  File "$spack/lib/spack/spack/cmd/__init__.py", line 164, in parse_specs
    spec.concretize(tests=tests)  # implies normalize
  File "$spack/lib/spack/spack/spec.py", line 2564, in concretize
    self._old_concretize(tests)
  File "$spack/lib/spack/spack/spec.py", line 2363, in _old_concretize
    self._expand_virtual_packages(concretizer),
  File "$spack/lib/spack/spack/spec.py", line 2266, in _expand_virtual_packages
    candidates = concretizer.choose_virtual_or_external(spec)
  File "$spack/lib/spack/spack/concretize.py", line 143, in choose_virtual_or_external
    candidates = self._valid_virtuals_and_externals(spec)
  File "$spack/lib/spack/spack/concretize.py", line 98, in _valid_virtuals_and_externals
    candidates = spack.repo.path.providers_for(spec)
  File "$spack/lib/spack/spack/repo.py", line 92, in converter
    return function(self, spec_like, *args, **kwargs)
  File "$spack/lib/spack/spack/repo.py", line 593, in providers_for
    providers = self.provider_index.providers_for(vpkg_spec)
  File "$spack/lib/spack/spack/repo.py", line 577, in provider_index
    self._provider_index.merge(repo.provider_index)
  File "$spack/lib/spack/spack/repo.py", line 991, in provider_index
    return self.index['providers']
  File "$spack/lib/spack/spack/repo.py", line 409, in __getitem__
    self._build_all_indexes()
  File "$spack/lib/spack/spack/repo.py", line 424, in _build_all_indexes
    self.indexes[name] = self._build_index(name, indexer)
  File "$spack/lib/spack/spack/repo.py", line 450, in _build_index
    indexer.read(old) if old else indexer.create()
  File "$spack/lib/spack/spack/repo.py", line 332, in read
    self.index = spack.provider_index.ProviderIndex.from_json(stream)
  File "$spack/lib/spack/spack/provider_index.py", line 294, in from_json
    lambda vpkg, plist: (
  File "$spack/lib/spack/spack/provider_index.py", line 323, in _transform
    for name, mappings in providers.items())
  File "$spack/lib/spack/spack/provider_index.py", line 323, in <genexpr>
    for name, mappings in providers.items())
  File "$spack/lib/spack/spack/provider_index.py", line 321, in <listcomp>
    [transform_fun(vpkg, pset) for vpkg, pset in mapiter(mappings)]
  File "$spack/lib/spack/spack/provider_index.py", line 295, in <lambda>
    spack.spec.Spec.from_node_dict(vpkg),
  File "$spack/lib/spack/spack/spec.py", line 1834, in from_node_dict
    spec.namespace = node.get('namespace', None)
AttributeError: 'str' object has no attribute 'get'
```
In PR #25890 I've put the v6-generation misc_cache into a new misc_cache subdirectory, and the v6 db in a new path.

--- a/lib/spack/spack/environment.py
+++ b/lib/spack/spack/environment.py
@@ -1743,6 +1743,8 @@ class Environment(object):
         specs_by_hash = {}
         for build_hash, node_dict in json_specs_by_hash.items():
             spec = Spec.from_node_dict(node_dict)
+            # Build hash is stored as a key, but not as part of the node dict
+            # To ensure build hashes are not recomputed, we reattach here
             setattr(spec, ht.build_hash.attr, build_hash)
             specs_by_hash[build_hash] = spec
 

--- a/lib/spack/spack/test/env.py
+++ b/lib/spack/spack/test/env.py
@@ -9,7 +9,7 @@ import spack.environment as ev
 import spack.spec
 
 
-def test_hash_change_no_rehash_concrete(tmpdir, mock_packages):
+def test_hash_change_no_rehash_concrete(tmpdir, mock_packages, config):
     # create an environment
     env_path = tmpdir.mkdir('env_dir').strpath
     env = ev.Environment(env_path)

--- a/lib/spack/spack/test/env.py
+++ b/lib/spack/spack/test/env.py
@@ -11,7 +11,8 @@ import spack.spec
 
 def test_hash_change_no_rehash_concrete(tmpdir, mock_packages):
     # create an environment
-    env = ev.Environment(tmpdir.mkdir('env_dir'))
+    env_path = tmpdir.mkdir('env_dir').strpath
+    env = ev.Environment(env_path)
     env.write()
 
     # add a spec with a rewritten build hash
@@ -29,9 +30,9 @@ def test_hash_change_no_rehash_concrete(tmpdir, mock_packages):
     env.write()
 
     # Read environment
-    read_in = ev.Environment(tmpdir.join('env_dir'))
+    read_in = ev.Environment(env_path)
 
-    # Ensure read hashes are used (new hash seen)
+    # Ensure read hashes are used (rewritten hash seen on read)
     assert read_in.concretized_order
     assert read_in.concretized_order[0] in read_in.specs_by_hash
     assert read_in.specs_by_hash[read_in.concretized_order[0]]._build_hash == new_hash

